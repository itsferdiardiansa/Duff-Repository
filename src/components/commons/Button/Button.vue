<template>
  <button 
    :class="customClass" 
    :disabled="disabled"
    @click="handleClick"
  >
    <slot>{{ label }}</slot>
  </button>
</template>
<script>
import { computed } from 'vue'

export default {
  props: {
    label: {
      type: String,
      default: '',
    },
    variant: {
      type: String,
      default: '',
    },
    size: {
      type: String,
      default: 'sm',
    },
    onClick: {
      type: Function,
      default: () => { },
    },
    textBold: {
      type: Boolean,
      default: false
    },
    disabled: {
      type: Boolean,
      default: false
    },
    rounded: {
      type: Boolean,
      default: true
    }
  },
  setup(props) {
    const customClass = computed(() => {
      let { variant, size, textBold, disabled, rounded } = props
      let _class = []

      _class.push(size)

      if(variant) _class.push(variant)
      
      if(!rounded) _class.push('no-rounded')

      if(textBold) _class.push('bold')

      if(disabled) _class.push('disabled')
      
      return _class.join(' ').trim()
    })

    const handleClick = () => {
      props.onClick()
    }

    return {
      customClass,
      handleClick,
    }
  },
}
</script>
<style lang="scss" scoped>
@import './variant.scss';
@import './size.scss';

button {
  @apply text-gray-50 focus:outline-none rounded py-2 px-3 transition duration-300 flex items-center justify-center;
  @apply bg-gray-200 hover:bg-gray-300 text-gray-800;

  @include variant;

  @include size;

  &.no-rounded {
    @apply rounded-none;
  }

  &.bold {
    @apply font-bold;
  }

  &.disabled {
    @apply disabled:opacity-70 pointer-events-none;
  }

}
</style>
