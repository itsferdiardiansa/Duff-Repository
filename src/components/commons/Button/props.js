export default {
  label: {
    type: String,
    default: '',
  },
  inverse: {
    type: Boolean,
    default: false
  },
  variant: {
    type: String,
    default: 'light',
    validator: function(variant) {
      return ~['primary', 'danger', 'warning', 'dark', 'success', 'light'].indexOf(variant)
    }
  },
  size: {
    type: String,
    default: 'base',
    validator: function(variant) {
      return ~['xs', 'sm', 'base', 'lg', 'xl'].indexOf(variant)
    }
  },
  icon: {
    type: Array,
    default: () => []
  },
  bold: {
    type: Boolean,
    default: false,
  },
  disabled: {
    type: Boolean,
    default: false,
  },
  rounded: {
    type: Boolean,
    default: true,
  },
  pill: {
    type: Boolean,
    default: false
  },
}