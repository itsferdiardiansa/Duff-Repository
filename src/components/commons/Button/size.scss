@mixin size {
  &.xs {
    @apply text-xs py-1.5 px-2;
  }

  &.sm {
    @apply text-sm p-2;
  }

  &.base {
    @apply text-base p-3;
  }

  &.lg {
    @apply text-lg p-4;
  }
}