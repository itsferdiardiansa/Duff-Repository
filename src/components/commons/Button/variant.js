export const buttonSize = {
  sm: 'py-1 px-2',
}