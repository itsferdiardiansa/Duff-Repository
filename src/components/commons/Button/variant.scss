@mixin variant {
  &.default {
    @apply bg-gray-200 hover:bg-gray-300 text-gray-800;
  }

  &.primary {
    @apply bg-blue-700 hover:bg-blue-800 text-gray-200;
  }

  &.danger {
    @apply bg-red-700 hover:bg-red-800 text-gray-200;
  }

  &.warning {
    @apply bg-yellow-500 hover:bg-yellow-600;
  }

  &.success {
    @apply bg-green-500 hover:bg-green-600 text-gray-200;
  }

  &.dark {
    @apply bg-gray-700 hover:bg-gray-800 text-gray-200;
  }
}