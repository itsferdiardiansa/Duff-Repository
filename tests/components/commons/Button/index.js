import { mount } from '@vue/test-utils'
import Button from '@common/Button'

describe('Button component', () => {
  let Component 

  beforeEach(async () => { 
    Component = mount(Button, {
      propsData: {
        onClick: jest.fn()
      }
    })

    await Component.setProps({
      label: '',
      variant: '',
      size: '',
      onClick: jest.fn(),
      textBold: false,
      disabled: false
    }) 
  })

  it('will render button correctly', () => {
    expect(Component.wrapper).toMatchSnapshot()
  })

  it('will have a content', async () => {
    await Component.setProps({ label: 'Your text here ...' })

    expect(Component.text()).toContain('Your text here ...')
  })

  it('will have a disabled button', async () => {
    await Component.setProps({ disabled: true })
    
    expect(Component.element.disabled).toBeTruthy()
  })

  it('will have a bold text', async () => {
    await Component.setProps({ textBold: true })

    expect(Component.attributes().class).toEqual('bold')
  })

  it('will not be rounded', async () => {
    await Component.setProps({ rounded: false })

    expect(Component.attributes().class).toEqual('no-rounded')
  })

  it('will have a size correctly', async () => {
    let sizes = ['sm', 'base', 'lg']

    for(let size of sizes) {
      await Component.setProps({ size })

      expect(Component.attributes().class).toContain(size)
    }
  })

  it('will have a background color correctly', async () => {
    let variants = ['primary', 'danger', 'warning', 'dark', 'success']

    for(let variant of variants) {
      await Component.setProps({ variant })

      expect(Component.attributes().class).toEqual(variant)
    }
  })

  it('will clicked twice', async () => {
    for(let i of [1,2]) {
      await Component.vm.onClick()
    }

    expect(Component.vm.onClick).toHaveBeenCalledTimes(2)
  })
})