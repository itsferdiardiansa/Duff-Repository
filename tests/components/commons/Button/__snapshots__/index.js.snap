// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component will render button correctly 1`] = `undefined`;
